### ISSUE TITLE
ridiculously low benchmarking results with nfIO

### ISSUE THREAD

**Issue by harendra-kumar** [vincenthz/hs-gauge#76]
Refer to [this line of code](https://github.com/composewell/streaming-benchmarks/blob/46f7db9de8935efc47cfef83f31d33b374c5144a/Benchmarks/Common.hs#L26). Reproduced below for easy reference:

```
benchIO name f n = bench name $ nfIO $ f n >>= return
```

If I remove the bind in this operation and just keep `f n`, some of the streaming libraries (conduit for example) being benchmarked by [streaming-benchmarks package](https://github.com/composewell/streaming-benchmarks) show ridiculously low results (in nanoseconds for a million operations). To reproduce, just remove that bind in that line and run this benchmark with and without it:

```
./run.sh elimination/toList/conduit
```

Clearly there is something wrong with that. Perhaps this function just gets optimized out because the compiler thinks the value is not being used? Or something else?

I did not investigate it much, I am hoping that someone knows what's going on and can figure this out quickly.

Without knowing this the results will remain unpredictable and benchmarking is of little use unless it is predictable. If the bind is really required then we can perhaps do that inside a wrapper so that the results are always predictable for users.

**Comment by harendra-kumar** [comment]
Even the ``return`` does not help in some cases. The issue maybe because the operation being benchmarked does not perform a real IO, it is more like a pure operation just using the IO monad. To avoid that I generated a random number and fed it to the operation being benchmarked. That made all such issues go away. We can either add this in the documentation or have some automatic workaround for this if possible.

**Comment by harendra-kumar** [comment]
This should be fixed by #84. Much thanks to @rubenpieters .
